testsuite/classlibrary/TestSynthDesc.sc

```diff
@@ -0,0 +1,88 @@
+TestSynthDesc : UnitTest {
+	test_SynthDesc_read_restores_metadata_in_desc_and_def {
+		var package = this.setupMetadataTests;
+		var synthdesc;
+
+		if(package.error.notNil) {
+			this.failed(thisMethod, "Error writing SynthDef: '%'".format(package.error.errorString));
+			^this.cleanupMetadataTests(package);
+		};
+
+		synthdesc = SynthDesc.read(package.path, keepDefs: true)[\test_md];
+		if(synthdesc.isNil) {
+			this.failed(thisMethod, "SynthDesc was not read from disk");
+			^this.cleanupMetadataTests(package);
+		};
+
+		this.checkMetadata(synthdesc, package.metadata, SynthDesc);
+
+		this.cleanupMetadataTests(package);
+	}
+
+	test_SynthDescLib_read_restores_metadata_in_desc_and_def {
+		var package = this.setupMetadataTests;
+		var synthdesc;
+
+		if(package.error.notNil) {
+			this.failed(thisMethod, "Error writing SynthDef: '%'".format(package.error.errorString));
+			^this.cleanupMetadataTests(package);
+		};
+
+		synthdesc = SynthDescLib.read(package.dir +/+ "*.scsyndef").global;
+		synthdesc = synthdesc[\test_md];
+		if(synthdesc.isNil) {
+			this.failed(thisMethod, "SynthDesc was not read from disk");
+			^this.cleanupMetadataTests(package);
+		};
+
+		this.checkMetadata(synthdesc, package.metadata, SynthDescLib);
+
+		this.cleanupMetadataTests(package);
+	}
+
+	setupMetadataTests {
+		var package = IdentityDictionary.new.know_(true);
+		package.dir = PathName.tmp;
+		package.path = package.dir +/+ "test_md.scsyndef";
+		package.mdPath = package.dir +/+ "test_md.txarcmeta";
+		package.metadata = (a: 1);
+		package.original = SynthDef(\test_md, {
+			Out.ar(0, DC.ar(0))
+		}).metadata_(package.metadata);
+
+		package.oldMdPlugin = SynthDesc.mdPlugin;
+		SynthDesc.mdPlugin = TextArchiveMDPlugin;
+
+		try {
+			package.original.writeDefFile(package.dir);
```

Comment:
what could cause an error here? maybe better to just let it throw?

Reply:
```
// user accidentally set PathName.tmp incorrectly
// we do no validation on a user's setting, so...
PathName.tmp = "/this/folder/doesnt/exist";

SynthDef(\abc, { Out.ar(0, SinOsc.ar(440)) }).writeDefFile(PathName.tmp);

ERROR: Primitive '_FilePutString' failed.
```

My thinking here was that we don't have total control over the user's environment, and we don't know which users will be running the test suite for which reasons. So I thought it would be better to have an environment-configuration issue reported as a test failure, allowing the test suite to continue. UnitTest doesn't wrap tests in error handlers -- so, any errors which individual tests don't catch will cause the entire test suite to die. That seems rude to me.

Also, though it seems unlikely, I suppose it isn't completely impossible for the file to be created but some error to occur while writing contents. In that case, the test should delete the file (but it can't, if there's no error handler).

If it's a general principle that we *do* want thrown errors to halt the entire test suite, then OK, I'll change it. But I would like a convincing reason to be given for that. IMO it would be better to log errors but not let them be fatal. (It would be even better if UnitTest did it for us, perhaps like this...)

```
	runTestMethod { | method, report = true |
		this.class.forkIfNeeded {
			try {
				this.setUp;
				currentMethod = method;
				this.perform(method.name);
			} { |error|
				this.failed(currentMethod, "Error thrown during test: '%'".format(error.errorString), report);
			};
			try {
				this.tearDown;
			} { |error|
				this.failed(currentMethod, "Error thrown during test cleanup: '%'".format(error.errorString), report);
			};
			if(report) { this.class.report };
		}
	}
```

Reply:
> My thinking here was that we don't have total control over the user's environment, and we don't know which users will be running the test suite for which reasons. So I thought it would be better to have an environment-configuration issue reported as a test failure, allowing the test suite to continue.

to me it's almost implied that with any software the test suite should be run in an unmodified environment. this means no additional configuration, add-ons, overrides, etc.; in fact i think we should probably document that. if we get into the defensive habit, we'll probably end up adding a lot of code, much of it boilerplate, for little gain.

> UnitTest doesn't wrap tests in error handlers -- so, any errors which individual tests don't catch will cause the entire test suite to die. That seems rude to me.

i agree, BUT i would rather solve that by catching errors in the test runner itself than by trying to anticipate every possible throw mid-test. that would again save a lot of boilerplate code, and also help to distinguish between intended and unintended errors, as well as intended and unintended test failures, which are very useful distinctions. so, my suggestion would be to let unintended errors bubble up, and trust that eventually it won't halt the entire test suite. in that light, handwritten error handling code is adding technical debt. if you feel strongly that it should remain, then i would suggest adding a comment with a TODO, XXX, FIXME, or other similar greppable marker that marks it as technical debt so that it can be found and removed later.

Reply:
OK, I've removed the error handling.

There are a couple of other meta-conversations about the test suite that I'll move out of this comment thread.